docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,190 @@
+   unscheduled resources). This value is the focus of this proposal. See [below](#api-changes) for
+   more details.
+4. **System Reserved** (proposed) - Compute resources reserved for system components such as the docker daemon, kubelet, kube proxy, etc.
+
+### API changes
+
+#### Allocatable
+
+Add `Allocatable` (4) to
+[`NodeStatus`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus):
+
+```
+type NodeStatus struct {
+  ...
+  // Allocatable represents schedulable resources of a node.
+  Allocatable ResourceList `json:"allocatable,omitempty"`
+  ...
+}
+```
+
+Allocatable will be computed by the Kubelet (unless explicitly set, see [below](#allocatable-flag))
+and reported periodically to the API server. It is defined to be:
+
+```
+   [Allocatable] = [Node Capacity] - ( [Reserved] + [System Reserved] + [Kernel usage (smoothed)] )
+```
+
+The scheduler will use `Allocatable` in place of `Capacity` when scheduling pods, and the Kubelet
+will use it when performing admission checks.
+
+- Since kernel usage can fluctuate and is out of kubernetes control, `Allocatable` may vary over
+  time.
+
+#### System Reserved
+
+`SystemReserved` is the parameter specifying resources reserved for system components (4). It is
+provided as a command-line flag to the Kubelet at startup, and therefore cannot be changed during
+normal Kubelet operation (this will change in the [future](#future-work)).
+
+The flag will be specified as a serialized `ResourceList`, with resources defined by the API
+`ResourecName` and values specified in `resource.Quantity` format, e.g.:
+
+```
+--system-reserved=cpu=500m,memory=5Mi
+```
+
+Initially we will only support CPU and memory, but will eventually support more resources. See
+[#16889](https://github.com/kubernetes/kubernetes/pull/16889) for disk accounting.
+
+If SystemReserved is not set it defaults to 0. Under this condition `Allocatable == Capacity`, and
+the system behavior is equivalent to the 1.1 behavior with scheduling based on Capacity.
+
+#### Reserved
+
+In the initial implementation, `Reserved` will be functionally equivalent to
+[`SystemReserved`](#system-reserved), but with a different semantic meaning. While system reserved
+designates resources set asside for system components, Reserved designates resources set asside for
+non-system components running outside of Kubernetes (currently this is reported as all the processes
+lumped together in the `/system` raw container).
+
+#### Allocatable flag
+
+If the `--allocatable` flag is set on the kubelet, then its value overrides the
+[`Allocatable`](#allocatable) calculation and is always reported instead. Initially this value can
+only be set at Kubelet start-up time, but will [eventually](#future-work) be converted to an
+updatable Config API (per-kubelet) parameter. Its value is specified using the same format as the
+[system reserved](#system-reserved) flag.
```

Comment:
My understanding is that there will be a bootstrapping config for kubelet on the local disk, which will then provide hints for the kubelet to register with a master and obtain its real-config data. Configuration can change dynamically and the kubelet might have to restart to adapt to that. Your argument is valid. Its just that coming up with a default behavior is difficult unless we do extensive benchmarking across different platforms.

Through this proposal, I feel we are adding cluster deployment functionality into kubelet.

If we look at system reserved from an admin perspective, the node can be configured to have limits on how much resources all the system daemons can consume. For example, a top level cgroup (`/system.slice`) could contain all the system daemons. Similarly, resources can be reserved for user login sessions. The remaining is what is available to the kubelet. Kubelet would then only care about the resources that are available for scheduling, and possibly its resource limits, if any.

I see this proposal as a means to communicate that node configuration to kubelet. Ideally, kubelet should not be attempting to partition the node. Cluster admins should do that. In such a world, there should not be a need for `reserved` and `system reserved` config params. There will be a cap on how much system daemons can use and so kubelet would only care about `capacity` which will not be equal to the overall node capacity. The config plane can remain outside of kubelet's scope.